### PR TITLE
feat: Add redirection support via raising subclasses of HTTPStatus

### LINKED ISSUE #406: Add a helper for redirects
Flask, Pyramid, Django, etc. all include helpers for redirects. It would be nice if falcon had something similar, especially if it could be used from hooks.

Currently, I'm implementing a redirect in the same before hook I do validation from (e.g. returning `404` if an object doesn't match a route), but it seems like the responder methods on the resource must check the status code after the hook so as not to overwrite the redirect.

### PR DESCRIPTION
**Pull request by kgriffs**
Define a set of HTTPStatus subclasses that can be raised to perform
various types of HTTP redirects. This should avoid the problem of
hooks and responder methods possibly overriding the redirect. Raising
an instance of one of these classes and will short-circuit request
processing similar to raising an instance of HTTPError.

Specifically, if raised in a before hook, it will skip any remaining
hooks and the responder method, but will not skip any process_response
middleware methods. If raised within a responder, it will skip the
rest of the responder and all after hooks. If raised in an after hook,
it would skip remaining after hooks but not middleware methods. And
finally, if raised within a middleware method, execution would perceive
as described at the bottom of [1].

The above behavior is inherited from HTTPStatus and so is not re-tested
in the subclasses.

[1]: https://falcon.readthedocs.org/en/stable/api/middleware.html

Closes #406
Supersedes #486

### CODE REVIEW

**Comment by kgriffs** [comment]
Reviewer: Please pay particular attention to docstrings, class and variable names, etc. to ensure everything is correct.

File: falcon/redirects.py [hunk 1]
```diff
@@ -0,0 +1,148 @@
...
+    """303 See Other.
+
+    The 303 (See Other) status code indicates that the server is
+    redirecting the user agent to a *different* resource, as indicated
+    by a URI in the Location header field, which is intended to provide
+    an indirect response to the original request.
+
+    A 303 response to a GET request indicates that the origin server
+    does not have a representation of the target resource that can be
+    transferred by the server over HTTP.  However, the Location field
```
Comment (jmvrbanac) [inline falcon/redirects.py]: You might clarify that this is a header field like you did earlier.

File: falcon/redirects.py [hunk 2]
```diff
@@ -0,0 +1,148 @@
...
+    value refers to a resource that is descriptive of the target
+    resource, such that making a retrieval request on that other
+    resource might result in a representation that is useful to
+    recipients without implying that it represents the original target
+    resource.
```
Comment (jmvrbanac) [inline falcon/redirects.py]: That is one long sentence. I had to re-read it a couple times to fully understand the point. If people have a problem understanding this, you might consider rephrasing or breaking this sentence up.

**Comment by jmvrbanac** [comment]
Other than my two comments, this LGTM.

**Comment by kgriffs** [comment]
@jmvrbanac IIRC the text for those docstrings came from the RFCs. Let me see if I can improve the prose per your comments.

**Comment by jmvrbanac** [comment]
:+1: LGTM

**Comment by kgriffs** [comment]
@rouge8 We are getting ready to merge this. Does this approach look good?

**Comment by rouge8** [comment]
Awesome! :+1: Ran my project's tests using this branch (and the minor code changes necessary). API worked as expected and the tests passed. Admittedly, I only used the 301, but this looks good!

**Comment by guoyucheng** [comment]
when the redirectspy can merge to the master?  I need redirects :)

**Pull request merged**